core/src/main/scala/kafka/server/AbstractAsyncFetcher.scala

```diff
@@ -0,0 +1,669 @@
+import org.apache.kafka.common.{InvalidRecordException, TopicPartition}
+
+import java.nio.ByteBuffer
+import java.util.Optional
+import java.util.concurrent.TimeUnit
+import java.util.function.Consumer
+import scala.collection.JavaConverters._
+import scala.collection.{Map, Seq, Set, mutable}
+import scala.math.min
+
+
+abstract class AbstractAsyncFetcher(clientId: String,
+                                    val sourceBroker: BrokerEndPoint,
+                                    failedPartitions: FailedPartitions,
+                                    fetchBackOffMs: Int = 0) extends Logging {
+
+  type FetchData = FetchResponse.PartitionData[Records]
+  type EpochData = OffsetsForLeaderEpochRequest.PartitionData
+
+  private val partitionStates = new PartitionStates[PartitionFetchState]
+
+  private val metricId = ClientIdAndBroker(clientId, sourceBroker.host, sourceBroker.port)
+  val fetcherStats = new AsyncFetcherStats(metricId)
+  val fetcherLagStats = new AsyncFetcherLagStats(metricId)
+
+  // process fetched data
+  protected def processPartitionData(topicPartition: TopicPartition,
+                                     fetchOffset: Long,
+                                     partitionData: FetchData): Option[LogAppendInfo]
+
+  protected def truncate(topicPartition: TopicPartition, truncationState: OffsetTruncationState): Unit
+
+  protected def truncateFullyAndStartAt(topicPartition: TopicPartition, offset: Long): Unit
+
+  protected def buildFetch(partitionMap: Map[TopicPartition, PartitionFetchState]): ResultWithPartitions[Option[FetchRequest.Builder]]
+
+  protected def latestEpoch(topicPartition: TopicPartition): Option[Int]
+
+  protected def logEndOffset(topicPartition: TopicPartition): Long
+
+  protected def endOffsetForEpoch(topicPartition: TopicPartition, epoch: Int): Option[OffsetAndEpoch]
+
+  protected def fetchEpochEndOffsets(partitions: Map[TopicPartition, EpochData]): Map[TopicPartition, EpochEndOffset]
+
+  protected def fetchFromLeader(fetchRequest: FetchRequest.Builder): Seq[(TopicPartition, FetchData)]
+
+  protected def fetchEarliestOffsetFromLeader(topicPartition: TopicPartition, currentLeaderEpoch: Int): Long
+
+  protected def fetchLatestOffsetFromLeader(topicPartition: TopicPartition, currentLeaderEpoch: Int): Long
+
+  protected def isOffsetForLeaderEpochSupported: Boolean
+
+
+  /**
+   * maybeFetch returns whether the fetching logic should back off
+   */
+  private def maybeFetch(): Boolean = {
+    val (fetchStates, fetchRequestOpt) = {
+      val fetchStates = partitionStates.partitionStateMap.asScala
+      val ResultWithPartitions(fetchRequestOpt, partitionsWithError) = buildFetch(fetchStates)
+
+      handlePartitionsWithErrors(partitionsWithError, "maybeFetch")
+
+      (fetchStates, fetchRequestOpt)
+    }
+    if (fetchRequestOpt.isEmpty) {
+      trace(s"There are no active partitions. Back off for $fetchBackOffMs ms before sending a fetch request")
```

Comment:
In `AbstractFetcherThread` when this case is hit, where there is nothing to do and the fetch is delayed (from `handlePartitionWithErrors`), maybeFetch waits for `fetchBackOffMs`.

Now that it is up to the caller of `maybeFetch` to wait, how is it going to know how long to wait? Does it just happen to know this backoff configuration value because the caller of `maybeFetch` is also the constructor of this AbstractAsyncFetcher? Or does it just not matter because subsequent calls will do nothing until the delay time has elapsed?

Actually, who calls `maybeFetch()`? That method is private and nothing currently references it. I assume the caller is coming in the next changeset?

Reply:
The caller will still be in this class and it will come in the next PR. If the caller sees backoff is needed, it will schedule a future event with a delay of the fetchBackOffMs, and thus the executing thread can be blocked up to that long.

Reply:
OK cool thank you for clarifying.